paddle/operators/detail/grpc_client.h

```diff
@@ -0,0 +1,269 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <time.h>
+#include <chrono>
+#include <ctime>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "paddle/framework/data_type.h"
+#include "paddle/framework/lod_tensor.h"
+#include "paddle/framework/scope.h"
+#include "paddle/framework/selected_rows.h"
+#include "paddle/operators/detail/simple_block_queue.h"
+
+#include "paddle/operators/detail/send_recv.grpc.pb.h"
+#include "paddle/operators/detail/send_recv.pb.h"
+
+#include <grpc++/grpc++.h>
+#include <grpc/support/log.h>
+
+namespace paddle {
+namespace operators {
+namespace detail {
+
+struct VarHandle {
+  std::string ep;
+  std::string name;
+  std::string String() const {
+    std::ostringstream s;
+    s << "name:" << name << " ep:" << ep;
+    return s.str();
+  }
+};
+
+struct SendStatus {
+  std::string error;
+  std::chrono::system_clock::time_point start;
+  std::chrono::system_clock::time_point end;
+  VarHandle var;
+
+  std::string String() const {
+    std::time_t t0 = std::chrono::system_clock::to_time_t(start);
+    std::time_t t1 = std::chrono::system_clock::to_time_t(end);
+    std::ostringstream s;
+    s << var.String() << " start_time:" << std::ctime(&t0)
+      << " end_time:" << std::ctime(&t1);
+    return s.str();
+  }
+};
+
+class ClientBase {
+ public:
+  virtual bool Call(const framework::Scope* scope, const std::string& name,
+                    int64_t time_out) = 0;
+  virtual bool ProcRet() = 0;
+  virtual grpc::Status* GetStatus() = 0;
+  virtual SendStatus GetSendStatus() = 0;
+
+  virtual ~ClientBase() = 0;
+};
+
+class SendClient : public ClientBase {
+ public:
+  SendClient(std::shared_ptr<grpc::CompletionQueue> cq,
+             std::shared_ptr<grpc::Channel> ch, std::string ep) {
+    cq_ = cq;
+    ch_ = ch;
+    ep_ = ep;
+    stub_ = sendrecv::SendRecvService::NewStub(ch);
+    context_.reset(new grpc::ClientContext());
```

Comment:
`context` should be created before every rpc send call.

Reply:
Done.Thanks.